library/Network/CloudSeeder/DSL.hs

```diff
@@ -9,72 +10,125 @@ module Network.CloudSeeder.DSL
+  }
 
 makeFields ''DeploymentConfiguration
 makeFields ''StackConfiguration
+makeFields ''HookContext
+
+class Monad m => CanSetConstant m where
+  param :: T.Text -> T.Text -> m ()
 
-deployment :: Monad m => T.Text -> StateT DeploymentConfiguration m a -> m DeploymentConfiguration
+instance Monad m => CanSetConstant (CreateT m) where
+  param key val = CreateT $ modify (M.insert key val)
+
+instance (Monad m, HasParameterSources s (S.Set (T.Text, ParameterSource))) => CanSetConstant (StateT s m) where
+  param key val = paramSource key (Constant val)
+
+deployment :: Monad m => T.Text -> StateT (DeploymentConfiguration m) m a -> m (DeploymentConfiguration m)
 deployment name' x =
   let config = DeploymentConfiguration name' [] [] []
   in execStateT x config
 
-stack_ :: Monad m => T.Text -> StateT DeploymentConfiguration m ()
+stack_ :: Monad m => T.Text -> StateT (DeploymentConfiguration m) m ()
 stack_ name' = stack name' $ return ()
 
-stack :: Monad m => T.Text -> StateT StackConfiguration m a -> StateT DeploymentConfiguration m ()
+stack :: Monad m => T.Text -> StateT (StackConfiguration m) m a -> StateT (DeploymentConfiguration m) m ()
 stack name' x = do
-  let stackConfig = StackConfiguration name' [] [] False
+  let stackConfig = StackConfiguration name' [] [] False []
   stackConfig' <- lift $ execStateT x stackConfig
   stacks %= (++ [stackConfig'])
 
 global :: (Monad m, HasGlobalStack a Bool) => StateT a m ()
 global = globalStack .= True
 
-environment :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource))) => [T.Text] -> StateT a m ()
+environment :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource)))
+            => [T.Text] -> StateT a m ()
 environment = mapM_ (`paramSource` Env)
 
-paramSource :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource)) ) => T.Text -> ParameterSource -> StateT a m ()
+paramSource :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource)))
+            => T.Text -> ParameterSource -> StateT a m ()
 paramSource pName source = parameterSources %= S.insert (pName, source)
 
-flag :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource))) => T.Text -> StateT a m ()
+flag :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource)))
+     => T.Text -> StateT a m ()
 flag pName = paramSource pName Flag
 
-param :: (Monad m, HasParameterSources a (S.Set (T.Text, ParameterSource)))
-      => T.Text -> T.Text -> StateT a m ()
-param key val = paramSource key (Constant val)
-
-tags :: (Monad m, HasTagSet a (S.Set (T.Text, T.Text))) => [(T.Text, T.Text)] -> StateT a m ()
+tags :: (Monad m, HasTagSet a (S.Set (T.Text, T.Text)))
+     => [(T.Text, T.Text)] -> StateT a m ()
 tags ts = tagSet %= (<> S.fromList ts)
+
+onCreate :: (Monad m, HasHooksCreate s [CreateT m ()])
+         => CreateT m () -> StateT s m ()
+onCreate action = hooksCreate %= (++ [action])
+
+databasePassword :: (MonadCloud m) => CreateT m ()
+databasePassword = do
```

Comment:
Does it make sense to:

  1. Hardcode the path within the bucket where this is stored…
  2. …and hardcode the size of the secret?

I think it would be nicer and more explicit to accept those as arguments to this function. Otherwise, it looks good.